File: packages/react-native-sdk/src/websocket-client.ts
import { ChatEvent, EventHandler, FluxyChatConfig } from './types';

export type ConnectionStatus = 'connecting' | 'connected' | 'disconnected' | 'reconnecting';

export class WebSocketClient {
  private wsUrl: string;
  private projectId: string;
  private token: string;
  private ws: WebSocket | null = null;
  private eventHandlers: Map<string, Set<EventHandler>> = new Map();
  private statusHandlers: Set<(status: ConnectionStatus) => void> = new Set();
  private reconnectAttempts = 0;
  private maxReconnectAttempts = 10;
  private reconnectTimer: ReturnType<typeof setTimeout> | null = null;
  private pingTimer: ReturnType<typeof setInterval> | null = null;
  private _status: ConnectionStatus = 'disconnected';
  private debug: boolean;

  constructor(config: FluxyChatConfig) {
    this.wsUrl = config.wsUrl.replace(/\/$/, '');
    this.projectId = config.projectId;
    this.token = config.token;
    this.debug = config.debug || false;
  }

  get status(): ConnectionStatus {
    return this._status;
  }

  connect(roomId: string): void {
    if (this.ws?.readyState === WebSocket.OPEN) {
      this.disconnect();
    }

    this.setStatus('connecting');
    const url = `${this.wsUrl}/ws?roomId=${roomId}&token=${this.token}&projectId=${this.projectId}`;

    try {
      this.ws = new WebSocket(url);

      this.ws.onopen = () => {
        this.log('WebSocket connected');
        this.setStatus('connected');
        this.reconnectAttempts = 0;
        this.startPing();
      };

      this.ws.onmessage = (event) => {
        try {
          const data = JSON.parse(event.data);
          this.handleMessage(data);
        } catch (e) {
          this.log('Failed to parse message', e);
        }
      };

      this.ws.onclose = (event) => {
        this.log('WebSocket closed', event.code, event.reason);
        this.stopPing();
        if (event.code !== 1000) {
          this.attemptReconnect(roomId);
        } else {
          this.setStatus('disconnected');
        }
      };

      this.ws.onerror = (error) => {
        this.log('WebSocket error', error);
        this.ws?.close();
      };
    } catch (e) {
      this.log('Failed to create WebSocket', e);
      this.attemptReconnect(roomId);
    }
  }

  disconnect(): void {
    this.stopPing();
    if (this.reconnectTimer) {
      clearTimeout(this.reconnectTimer);
      this.reconnectTimer = null;
    }
    this.reconnectAttempts = this.maxReconnectAttempts;
    this.ws?.close(1000, 'Client disconnect');
    this.ws = null;
    this.setStatus('disconnected');
  }

  send(data: Record<string, unknown>): void {
    if (this.ws?.readyState === WebSocket.OPEN) {
      this.ws.send(JSON.stringify(data));
    } else {
      this.log('Cannot send, WebSocket not connected');
    }
  }

  sendTyping(roomId: string, typing: boolean): void {
    this.send({ type: 'typing', roomId, typing });
  }

  sendPresence(roomId: string, status: string): void {
    this.send({ type: 'presence', roomId, status });
  }

  sendReadReceipt(roomId: string, messageId: string): void {
    this.send({ type: 'read', roomId, messageId });
  }

  on(event: string, handler: EventHandler): () => void {
    if (!this.eventHandlers.has(event)) {
      this.eventHandlers.set(event, new Set());
    }
    this.eventHandlers.get(event)!.add(handler);
    return () => this.off(event, handler);
  }

  off(event: string, handler: EventHandler): void {
    this.eventHandlers.get(event)?.delete(handler);
  }

  onStatusChange(handler: (status: ConnectionStatus) => void): () => void {
    this.statusHandlers.add(handler);
    return () => this.statusHandlers.delete(handler);
  }

  private handleMessage(data: ChatEvent): void {
    const handlers = this.eventHandlers.get(data.type);
    if (handlers) {
      handlers.forEach((handler) => handler(data));
    }
    const allHandlers = this.eventHandlers.get('*');
    if (allHandlers) {
      allHandlers.forEach((handler) => handler(data));
    }
  }

  private attemptReconnect(roomId: string): void {
    if (this.reconnectAttempts >= this.maxReconnectAttempts) {
      this.log('Max reconnect attempts reached');
      this.setStatus('disconnected');
      return;
    }

    this.setStatus('reconnecting');
    const delay = Math.min(1000 * Math.pow(2, this.reconnectAttempts), 30000);
    this.log(`Reconnecting in ${delay}ms (attempt ${this.reconnectAttempts + 1})`);

    this.reconnectTimer = setTimeout(() => {
      this.reconnectAttempts++;
      this.connect(roomId);
    }, delay);
  }

  private startPing(): void {
    this.pingTimer = setInterval(() => {
      this.send({ type: 'ping' });
    }, 30000);
  }

  private stopPing(): void {
    if (this.pingTimer) {
      clearInterval(this.pingTimer);
      this.pingTimer = null;
    }
  }

  private setStatus(status: ConnectionStatus): void {
    this._status = status;
    this.statusHandlers.forEach((handler) => handler(status));
  }

  private log(...args: unknown[]): void {
    if (this.debug) {
      console.log('[FluxyChat WS]', ...args);
    }
  }
}

File: packages/flutter-sdk/lib/src/websocket_client.dart
import 'dart:async';
import 'dart:convert';
import 'package:web_socket_channel/web_socket_channel.dart';
import 'models.dart';

typedef EventHandler = void Function(ChatEvent event);

class WebSocketClient {
  final String wsUrl;
  final String projectId;
  final String token;
  final bool debug;

  WebSocketChannel? _channel;
  final Map<String, Set<EventHandler>> _eventHandlers = {};
  final Set<void Function(ConnectionStatus)> _statusHandlers = {};
  ConnectionStatus _status = ConnectionStatus.disconnected;
  int _reconnectAttempts = 0;
  final int _maxReconnectAttempts = 10;
  Timer? _reconnectTimer;
  Timer? _pingTimer;
  String? _currentRoomId;

  WebSocketClient({
    required this.wsUrl,
    required this.projectId,
    required this.token,
    this.debug = false,
  });

  ConnectionStatus get status => _status;

  void connect(String roomId) {
    _currentRoomId = roomId;
    _setStatus(ConnectionStatus.connecting);

    final url = '$wsUrl/ws?roomId=$roomId&token=$token&projectId=$projectId';
    _channel = WebSocketChannel.connect(Uri.parse(url));

    _channel!.stream.listen(
      (data) {
        try {
          final json = jsonDecode(data as String);
          _handleMessage(json);
        } catch (e) {
          _log('Failed to parse message: $e');
        }
      },
      onDone: () {
        _log('WebSocket closed');
        _stopPing();
        if (_reconnectAttempts < _maxReconnectAttempts) {
          _attemptReconnect();
        } else {
          _setStatus(ConnectionStatus.disconnected);
        }
      },
      onError: (error) {
        _log('WebSocket error: $error');
        _channel?.sink.close();
      },
    );

    _setStatus(ConnectionStatus.connected);
    _reconnectAttempts = 0;
    _startPing();
  }

  void disconnect() {
    _stopPing();
    _reconnectTimer?.cancel();
    _reconnectAttempts = _maxReconnectAttempts;
    _channel?.sink.close(1000, 'Client disconnect');
    _channel = null;
    _setStatus(ConnectionStatus.disconnected);
  }

  void send(Map<String, dynamic> data) {
    if (_channel != null) {
      _channel!.sink.add(jsonEncode(data));
    } else {
      _log('Cannot send, WebSocket not connected');
    }
  }

  void sendTyping(String roomId, bool typing) {
    send({'type': 'typing', 'roomId': roomId, 'typing': typing});
  }

  void sendPresence(String roomId, String status) {
    send({'type': 'presence', 'roomId': roomId, 'status': status});
  }

  void sendReadReceipt(String roomId, String messageId) {
    send({'type': 'read', 'roomId': roomId, 'messageId': messageId});
  }

  void on(String event, EventHandler handler) {
    _eventHandlers.putIfAbsent(event, () => {}).add(handler);
  }

  void off(String event, EventHandler handler) {
    _eventHandlers[event]?.remove(handler);
  }

  void onStatusChange(void Function(ConnectionStatus) handler) {
    _statusHandlers.add(handler);
  }

  void _handleMessage(Map<String, dynamic> data) {
    final event = ChatEvent(
      type: data['type'],
      data: data['data'],
      roomId: data['roomId'],
      timestamp: DateTime.parse(data['timestamp'] ?? DateTime.now().toIso8601String()),
    );

    _eventHandlers[data['type']]?.forEach((handler) => handler(event));
    _eventHandlers['*']?.forEach((handler) => handler(event));
  }

  void _attemptReconnect() {
    _setStatus(ConnectionStatus.reconnecting);
    final delay = Duration(milliseconds: (1000 * (1 << _reconnectAttempts)).clamp(1000, 30000));
    _log('Reconnecting in ${delay.inMilliseconds}ms (attempt ${_reconnectAttempts + 1})');

    _reconnectTimer = Timer(delay, () {
      _reconnectAttempts++;
      if (_currentRoomId != null) connect(_currentRoomId!);
    });
  }

  void _startPing() {
    _pingTimer = Timer.periodic(const Duration(seconds: 30), (_) {
      send({'type': 'ping'});
    });
  }

  void _stopPing() {
    _pingTimer?.cancel();
    _pingTimer = null;
  }

  void _setStatus(ConnectionStatus status) {
    _status = status;
    _statusHandlers.forEach((handler) => handler(status));
  }

  void _log(String message) {
    if (debug) {
      print('[FluxyChat WS] $message');
    }
  }
}
